// apify/configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Configuration = void 0;
const core_1 = require("@crawlee/core");
const consts_1 = require("@apify/consts");
/**
 * `Configuration` is a value object holding the SDK configuration. We can use it in two ways:
 *
 * 1. When using `Actor` class, we can get the instance configuration via `sdk.config`
 *
 *    ```javascript
 *    import { Actor } from 'apify';
 *    import { BasicCrawler } from 'crawlee';
 *
 *    const sdk = new Actor({ token: '123' });
 *    console.log(sdk.config.get('token')); // '123'
 *
 *    const crawler = new BasicCrawler({
 *        // ... crawler options
 *    }, sdk.config);
 *    ```
 *
 * 2. To get the global configuration (singleton instance). It will respect the environment variables.
 *
 *    ```javascript
 *    import { BasicCrawler, Configuration } from 'crawlee';
 *
 *    // Get the global configuration
 *    const config = Configuration.getGlobalConfig();
 *    // Set the 'persistStateIntervalMillis' option
 *    // of global configuration to 30 seconds
 *    config.set('persistStateIntervalMillis', 30_000);
 *
 *    // No need to pass the configuration to the crawler,
 *    // as it's using the global configuration by default
 *    const crawler = new BasicCrawler();
 *    ```
 *
 * ## Supported Configuration Options
 *
 * Key | Environment Variable | Default Value
 * ---|---|---
 * `memoryMbytes` | `ACTOR_MEMORY_MBYTES` | -
 * `headless` | `APIFY_HEADLESS` | -
 * `persistStateIntervalMillis` | `APIFY_PERSIST_STATE_INTERVAL_MILLIS` | `60e3`
 * `token` | `APIFY_TOKEN` | -
 * `isAtHome` | `APIFY_IS_AT_HOME` | -
 * `defaultDatasetId` | `ACTOR_DEFAULT_DATASET_ID` | `'default'`
 * `defaultKeyValueStoreId` | `ACTOR_DEFAULT_KEY_VALUE_STORE_ID` | `'default'`
 * `defaultRequestQueueId` | `ACTOR_DEFAULT_REQUEST_QUEUE_ID` | `'default'`
 *
 * ## Advanced Configuration Options
 *
 * Key | Environment Variable | Default Value
 * ---|---|---
 * `actorEventsWsUrl` | `ACTOR_EVENTS_WEBSOCKET_URL` | -
 * `actorId` | `ACTOR_ID` | -
 * `actorRunId` | `ACTOR_RUN_ID` | -
 * `actorTaskId` | `ACTOR_TASK_ID` | -
 * `apiBaseUrl` | `APIFY_API_BASE_URL` | `'https://api.apify.com'`
 * `containerPort` | `ACTOR_WEB_SERVER_PORT` | `4321`
 * `containerUrl` | `ACTOR_WEB_SERVER_URL` | `'http://localhost:4321'`
 * `inputKey` | `ACTOR_INPUT_KEY` | `'INPUT'`
 * `metamorphAfterSleepMillis` | `APIFY_METAMORPH_AFTER_SLEEP_MILLIS` | `300e3`
 * `metaOrigin` | `APIFY_META_ORIGIN` | -
 * `proxyHostname` | `APIFY_PROXY_HOSTNAME` | `'proxy.apify.com'`
 * `proxyPassword` | `APIFY_PROXY_PASSWORD` | -
 * `proxyPort` | `APIFY_PROXY_PORT` | `8000`
 * `proxyStatusUrl` | `APIFY_PROXY_STATUS_URL` | `'http://proxy.apify.com'`
 * `userId` | `APIFY_USER_ID` | -
 * `userIsPaying` | `APIFY_USER_IS_PAYING` | -
 * `xvfb` | `APIFY_XVFB` | -
 * `standbyPort` | `ACTOR_STANDBY_PORT` | `4321`
 * `standbyUrl` | `ACTOR_STANDBY_URL` | -
 * `chromeExecutablePath` | `APIFY_CHROME_EXECUTABLE_PATH` | -
 * `defaultBrowserPath` | `APIFY_DEFAULT_BROWSER_PATH` | -
 */
class Configuration extends core_1.Configuration {
    /**
     * @inheritDoc
     */
    get(key, defaultValue) {
        return super.get(key, defaultValue);
    }
    /**
     * @inheritDoc
     */
    set(key, value) {
        super.set(key, value);
    }
    /**
     * @inheritDoc
     */
    static getGlobalConfig() {
        if (Configuration.storage.getStore()) {
            return Configuration.storage.getStore();
        }
        Configuration.globalConfig ?? (Configuration.globalConfig = new Configuration());
        return Configuration.globalConfig;
    }
    /**
     * Resets global configuration instance. The default instance holds configuration based on env vars,
     * if we want to change them, we need to first reset the global state. Used mainly for testing purposes.
     */
    static resetGlobalState() {
        delete this.globalConfig;
    }
}
exports.Configuration = Configuration;
// maps environment variables to config keys (e.g. `APIFY_MEMORY_MBYTES` to `memoryMbytes`)
Object.defineProperty(Configuration, "ENV_MAP", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: {
        // regular crawlee env vars are also supported
        ...core_1.Configuration.ENV_MAP,
        // support crawlee env vars prefixed with `APIFY_` too
        APIFY_AVAILABLE_MEMORY_RATIO: 'availableMemoryRatio',
        APIFY_PURGE_ON_START: 'purgeOnStart',
        APIFY_MEMORY_MBYTES: 'memoryMbytes',
        APIFY_DEFAULT_DATASET_ID: 'defaultDatasetId',
        APIFY_DEFAULT_KEY_VALUE_STORE_ID: 'defaultKeyValueStoreId',
        APIFY_DEFAULT_REQUEST_QUEUE_ID: 'defaultRequestQueueId',
        APIFY_INPUT_KEY: 'inputKey',
        APIFY_PERSIST_STATE_INTERVAL_MILLIS: 'persistStateIntervalMillis',
        APIFY_HEADLESS: 'headless',
        APIFY_XVFB: 'xvfb',
        APIFY_CHROME_EXECUTABLE_PATH: 'chromeExecutablePath',
        APIFY_DEFAULT_BROWSER_PATH: 'defaultBrowserPath',
        APIFY_DISABLE_BROWSER_SANDBOX: 'disableBrowserSandbox',
        // as well as apify specific ones
        APIFY_TOKEN: 'token',
        APIFY_METAMORPH_AFTER_SLEEP_MILLIS: 'metamorphAfterSleepMillis',
        APIFY_TEST_PERSIST_INTERVAL_MILLIS: 'persistStateIntervalMillis', // for BC, seems to be unused
        APIFY_ACTOR_EVENTS_WS_URL: 'actorEventsWsUrl',
        APIFY_ACTOR_ID: 'actorId',
        APIFY_API_BASE_URL: 'apiBaseUrl',
        APIFY_API_PUBLIC_BASE_URL: 'apiPublicBaseUrl',
        APIFY_IS_AT_HOME: 'isAtHome',
        APIFY_ACTOR_RUN_ID: 'actorRunId',
        APIFY_ACTOR_TASK_ID: 'actorTaskId',
        APIFY_CONTAINER_PORT: 'containerPort',
        APIFY_CONTAINER_URL: 'containerUrl',
        APIFY_USER_ID: 'userId',
        APIFY_USER_IS_PAYING: 'userIsPaying',
        APIFY_PROXY_HOSTNAME: 'proxyHostname',
        APIFY_PROXY_PASSWORD: 'proxyPassword',
        APIFY_PROXY_STATUS_URL: 'proxyStatusUrl',
        APIFY_PROXY_PORT: 'proxyPort',
        APIFY_INPUT_SECRETS_PRIVATE_KEY_FILE: 'inputSecretsPrivateKeyFile',
        APIFY_INPUT_SECRETS_PRIVATE_KEY_PASSPHRASE: 'inputSecretsPrivateKeyPassphrase',
        APIFY_META_ORIGIN: 'metaOrigin',
        // Actor env vars
        ACTOR_DEFAULT_DATASET_ID: 'defaultDatasetId',
        ACTOR_DEFAULT_KEY_VALUE_STORE_ID: 'defaultKeyValueStoreId',
        ACTOR_DEFAULT_REQUEST_QUEUE_ID: 'defaultRequestQueueId',
        ACTOR_EVENTS_WEBSOCKET_URL: 'actorEventsWsUrl',
        ACTOR_ID: 'actorId',
        ACTOR_INPUT_KEY: 'inputKey',
        ACTOR_MEMORY_MBYTES: 'memoryMbytes',
        ACTOR_RUN_ID: 'actorRunId',
        ACTOR_STANDBY_PORT: 'standbyPort',
        ACTOR_STANDBY_URL: 'standbyUrl',
        ACTOR_TASK_ID: 'actorTaskId',
        ACTOR_WEB_SERVER_PORT: 'containerPort',
        ACTOR_WEB_SERVER_URL: 'containerUrl',
        ACTOR_MAX_TOTAL_CHARGE_USD: 'maxTotalChargeUsd',
        ACTOR_TEST_PAY_PER_EVENT: 'testPayPerEvent',
        ACTOR_USE_CHARGING_LOG_DATASET: 'useChargingLogDataset',
    }
});
Object.defineProperty(Configuration, "INTEGER_VARS", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: [
        ...core_1.Configuration.INTEGER_VARS,
        'proxyPort',
        'containerPort',
        'metamorphAfterSleepMillis',
        'maxTotalChargeUsd',
    ]
});
Object.defineProperty(Configuration, "BOOLEAN_VARS", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: [
        ...core_1.Configuration.BOOLEAN_VARS,
        'isAtHome',
        'testPayPerEvent',
        'useChargingLogDataset',
    ]
});
Object.defineProperty(Configuration, "DEFAULTS", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: {
        ...core_1.Configuration.DEFAULTS,
        defaultKeyValueStoreId: consts_1.LOCAL_ACTOR_ENV_VARS[consts_1.ACTOR_ENV_VARS.DEFAULT_KEY_VALUE_STORE_ID],
        defaultDatasetId: consts_1.LOCAL_ACTOR_ENV_VARS[consts_1.ACTOR_ENV_VARS.DEFAULT_DATASET_ID],
        defaultRequestQueueId: consts_1.LOCAL_ACTOR_ENV_VARS[consts_1.ACTOR_ENV_VARS.DEFAULT_REQUEST_QUEUE_ID],
        inputKey: 'INPUT',
        apiBaseUrl: 'https://api.apify.com',
        apiPublicBaseUrl: 'https://api.apify.com',
        proxyStatusUrl: 'http://proxy.apify.com',
        proxyHostname: consts_1.LOCAL_APIFY_ENV_VARS[consts_1.APIFY_ENV_VARS.PROXY_HOSTNAME],
        proxyPort: +consts_1.LOCAL_APIFY_ENV_VARS[consts_1.APIFY_ENV_VARS.PROXY_PORT],
        containerPort: +consts_1.LOCAL_ACTOR_ENV_VARS[consts_1.ACTOR_ENV_VARS.WEB_SERVER_PORT],
        containerUrl: consts_1.LOCAL_ACTOR_ENV_VARS[consts_1.ACTOR_ENV_VARS.WEB_SERVER_URL],
        standbyPort: +consts_1.LOCAL_ACTOR_ENV_VARS[consts_1.ACTOR_ENV_VARS.STANDBY_PORT],
        metamorphAfterSleepMillis: 300e3,
        persistStateIntervalMillis: 60e3, // This value is mentioned in jsdoc in `events.js`, if you update it here, update it there too.
        testPayPerEvent: false,
        useChargingLogDataset: false,
    }
});
// monkey patch the core class so it respects the new options too
core_1.Configuration.getGlobalConfig = Configuration.getGlobalConfig;
// @ts-expect-error protected property
core_1.Configuration.ENV_MAP = Configuration.ENV_MAP;
// @ts-expect-error protected property
core_1.Configuration.INTEGER_VARS = Configuration.INTEGER_VARS;
// @ts-expect-error protected property
core_1.Configuration.BOOLEAN_VARS = Configuration.BOOLEAN_VARS;
// @ts-expect-error protected property
core_1.Configuration.DEFAULTS = Configuration.DEFAULTS;
